components/engine_rocks/src/raft_compaction_filter.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::raft_engine::{RAFT_LOG_GC_INDEXES, RAFT_LOG_GC_ON_COMPACTION};
+use crate::raw::{
+    new_compaction_filter_raw, CompactionFilter, CompactionFilterContext, CompactionFilterDecision,
+    CompactionFilterFactory, CompactionFilterValueType, DBCompactionFilter,
+};
+use std::ffi::CString;
+use tikv_util::debug;
+
+pub struct RaftLogCompactionFilterFactory {}
+
+impl RaftLogCompactionFilterFactory {}
+
+impl CompactionFilterFactory for RaftLogCompactionFilterFactory {
+    fn create_compaction_filter(
+        &self,
+        _context: &CompactionFilterContext,
+    ) -> *mut DBCompactionFilter {
+        let filter = Box::new(RaftLogCompactionFilter::new(
+            RAFT_LOG_GC_ON_COMPACTION.load(std::sync::atomic::Ordering::Acquire),
+        ));
+        let name = CString::new("raft_log_gc_compaction_filter").unwrap();
+        unsafe { new_compaction_filter_raw(name, filter) }
+    }
+}
+
+struct RaftLogCompactionFilter {
+    enabled: bool,
+    last_region: u64,
+    gc_point: u64,
+}
+
+impl RaftLogCompactionFilter {
+    fn new(enabled: bool) -> Self {
+        RaftLogCompactionFilter {
+            enabled,
+            // for region id will NEVER be 0
+            last_region: 0,
+            gc_point: 0,
+        }
+    }
+}
+
+impl CompactionFilter for RaftLogCompactionFilter {
+    fn featured_filter(
+        &mut self,
+        _level: usize,
+        key: &[u8],
+        _sequence: u64,
+        _value: &[u8],
+        value_type: CompactionFilterValueType,
+    ) -> CompactionFilterDecision {
+        if !self.enabled {
+            return CompactionFilterDecision::Keep;
+        }
+        if value_type != CompactionFilterValueType::Value {
+            return CompactionFilterDecision::Keep;
+        }
+        let (current_region, current_index) = match keys::decode_raft_log_key(key) {
+            Ok((rid, idx)) => (rid, idx),
+            Err(_) => {
+                debug!(
+                    "keep non raft log key when raft log gc compaction";
+                    "key"=>format!("{:?}",key),
+                );
+                return CompactionFilterDecision::Keep;
+            }
+        };
+
+        if current_region != self.last_region {
+            self.last_region = current_region;
+            let indexes = RAFT_LOG_GC_INDEXES.read().unwrap();
+            self.gc_point = indexes.get(&current_region).unwrap_or(&0).to_owned();
+        }
+
+        if current_index < self.gc_point {
+            return CompactionFilterDecision::Remove;
```

Comment:
It still leave tombstones. Use `Decision::RemoveAndSkipUntil` instead can avoid tombstones.